Blackboard json

christianhelm opened the pull request:
Support for blackboards defined in json. 
Can use own blackboard or inherited blackboard.
Supports key mapping.
https://www.wrike.com/open.htm?id=816946640

christianhelm committed:
Add test

christianhelm committed:
Read behavior blackboard from file

christianhelm committed:
Read plan blackboard from json

christianhelm committed:
Blackboard set function

christianhelm committed:
Add empty blackboard

christianhelm committed:
Inherit blackboard test

christianhelm committed:
Passing parameters in

christianhelm committed:
Add non-converting getter

christianhelm committed:
Pass parameters on termination

christianhelm committed:
Test for passing parameters on termination

christianhelm committed:
Add tests

christianhelm committed:
Update tests

christianhelm committed:
Flag tests

christianhelm committed:
Improve inherit test

christianhelm committed:
Pass in plan

christianhelm committed:
Default inherit true

christianhelm committed:
Merge branch 'rr_devel' into blackboard_json_2

christianhelm committed:
Fix termination

christianhelm committed:
Fix older tests

christianhelm committed:
Fix related code

christianhelm committed:
Merge branch 'rr_devel' into blackboard_json_2

christianhelm committed:
Fix comment

christianhelm committed:
Cleanup

Dale-Koenig commented on alica_tests/src/test/etc/plans/behaviours/TestParameterPassingBehaviour.beh:
```diff
@@ -10,5 +10,29 @@
   "requiresParameters": true,
   "runtimeCondition": null,
-  "variables": []
+  "variables": [],
```
Must manaully set relativeDirectory to `behaviours` at the moment or everything goes in the wrong place

Dale-Koenig commented on alica_tests/src/test/test_inherit_blackboard.cpp:
```diff
@@ -0,0 +1,38 @@
+    ae->getAlicaClock().sleep(alica::AlicaTime::milliseconds(100));
+    auto wm = dynamic_cast<alicaTests::TestWorldModel*>(ae->getWorldModel());
+    EXPECT_EQ(wm->passedParameters["masterKeyInBehavior"], 3);
+    EXPECT_EQ(wm->passedParameters["hasBehaviorKey"], 4);
```
I don't get how this test is for inherited blackboards

christianhelm replied:
It checks if a key that is only available through inheritance is available.
The second check tests if a key defined in a child is not available in the parent when they are connected by inheritance. I have added comments in the code that explain that.

Dale-Koenig replied:
I guess what I am confused about is the relationship between blackboards and the world model.  If this is a test for blackboards, why are we checking the world model?

christianhelm replied:
The blackboard itself is not accessible in the test code. The behaviors and plans read the blackboard and write into the world model. Then the test can check the world model. It's a pattern I found in several other tests and used again here.

Dale-Koenig commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -61,7 +62,7 @@ class BasicBehaviour : private RunnableObject
     const Variable* getVariable(const std::string& name) const { return _behaviour->getVariable(name); };
 
-    void terminate() { stop(); };
+    void terminate(RunningPlan* rp) { stop(rp); };
```
What is the point of this function?  Why not call stop directly?

christianhelm replied:
I'm not sure myself. I just left the naming/pattern as it was.

abhi1287 replied:
looks like this isn't used anymore and can be removed

Dale-Koenig commented on alica_engine/src/engine/blackboard/KeyMapping.cpp:
```diff
@@ -0,0 +1,44 @@
+
+void KeyMapping::setInput(const Blackboard* parent_bb, Blackboard* child_bb) const
+{
+    const auto lockedParentBb = LockedBlackboardRO(*parent_bb);
```
I guess we are guaranteed no thread will be locking child blackboard when we are doing this (or setOutput)?

christianhelm replied:
On set input: Parent might be locked, Child is guaranteed to be not locked.
On set output: Child is guaranteed to be not locked. Parent might be locked (e.g. if blackboard is used in state transition check) 
It can only lead to problems if the lock is kept forever, otherwise it will become available at some point

abhi1287 replied:
so child is not required to be locked in both the cases?

Dale-Koenig commented on alica_tests/src/test/etc/plans/TestParameterPassing.pml:
```diff
@@ -35,25 +35,154 @@
+            [
+              {
+                "parent" : "planOutputKey",
+                "child" : "behaviorInputKey"
```
I think it would be clearer to call these keys parent_key, child_key

Dale-Koenig commented on alica_engine/include/engine/modelmanagement/factories/BlackboardFactory.h:
```diff
@@ -0,0 +1,19 @@
+class BlackboardFactory : public Factory
+{
+public:
+    static std::unique_ptr<Blackboard> create(const YAML::Node& node);
```
Can we just return a shared_ptr here?  Returning a unique ptr then moving it into a shared ptr is kind of weird

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -42,7 +42,37 @@ void RunnableObject::stop()
+            const auto keyMapping = parentPlan->getKeyMapping(wrapperId);
+            auto terminateCall = [this, keyMapping, parentPlan = parentPlan]() {
+                assert(_blackboard);
+                keyMapping.setOutput(parentPlan->getBlackboard().get(), _blackboard.get());
```
Can we set the output mapping the same time we set the input mapping (when initializing the behavior/plan)?

christianhelm replied:
The output must be copied on termination, so I don't see a another point where this could be done.

Dale-Koenig replied:
I suppose the scheduling can only be done here, but the keyMapping lookup could have been cached from earlier to avoid going through all the wrappers etc again.  I suppose it would have to be cached from the scheduled init() call though, so maybe easier to just go through the wrappers again here.

Mostly just complaining about the duplicated code, so maybe the solution is to refactor out the code to find the appropriate keyMapping?

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -52,43 +82,39 @@ void RunnableObject::start(RunningPlan* rp)
-            doInit();
-        };
-        _engine->editScheduler().schedule(initCall);
+    if (rp->getParent()) {
```
I guess we could reduce a level of nesting by starting with
```
if (!rp->getParent) {
...
} else if(!_inheritBlackboard) {
```

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -52,43 +82,39 @@ void RunnableObject::start(RunningPlan* rp)
+            const auto keyMapping = parentPlan->getKeyMapping(wrapperId);
+            auto initCall = [this, keyMapping, parentPlan = parentPlan]() {
+                assert(_blackboard);
+                keyMapping.setInput(parentPlan->getBlackboard().get(), _blackboard.get());
```
I think we still need to reset child blackboard here.  The child may have protected parameters that have leftover values from previous runs.

I guess `clear()` needs to be replaced with `reset()` uch that it does not remove keys, but just resets everything to their default values

christianhelm replied:
I have changed the way blackboards are created. Now every call of a non-inherited blackboard generates a new instance. A clear/reset should not be necessary.

christianhelm committed:
Use new code generator

christianhelm committed:
Update generated conde ins supplementary

christianhelm committed:
Restructure nested if

christianhelm committed:
Correct relative directory

christianhelm committed:
Add comments

christianhelm committed:
Remove PlanAttachment include

christianhelm committed:
Rename parent/child

christianhelm committed:
Use blackboard blueprint

christianhelm committed:
Test launch file improvements
These modificiations are already used in other projects
Xterm config
Gtest filter

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -63,33 +101,27 @@ void RunnableObject::start(RunningPlan* rp)
+        BasicPlan* parentPlan = rp->getParent()->getBasicPlan();
+        initCall = [this, parentPlan]() {
             if (parentPlan) {
                 _blackboard = parentPlan->getBlackboard();
```
Does the check for parentPlan do anything here?  case with no parent should now be handled above right?

christianhelm committed:
Merge branch 'test_debug_launch' into blackboard_json_2

christianhelm committed:
Restructure if statements

christianhelm committed:
Refacturing to remove duplicate code

abhi1287 commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -19,27 +20,28 @@ class BasicPlan : private RunnableObject
     using RunnableObject::getTraceContext;
     using RunnableObject::getWorldModel;
     using RunnableObject::initExecuted;
+    using RunnableObject::setBlackboardBlueprint;
```
You can get rid of all these functions by simply passing (const BlackboardBlueprint *) in BasicPlan and BasicBehaviour constructor and storing it as const pointer as well in RunnableObject.  A nullptr would signal that no blueprint exists (inherit blackboard).

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -81,11 +81,11 @@ void BasicPlan::doTerminate()
     for (const State* state : plan->getStates()) {
         for (const ConfAbstractPlanWrapper* wrapper : state->getConfAbstractPlanWrappers()) {
-            _planAttachments.emplace(wrapper->getId(), planCreator.createPlanAttachment(wrapper->getId()));
+            _keyMappings.emplace(wrapper->getId(), wrapper->getKeyMapping());
```
Since the mappings are never going to change, keyMapping doesn't need to be duplicated every time a RunnableObject is created.  You can return a const pointer / reference from wrapper->getKeyMapping()  and just pass around the reference here.

abhi1287 commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -42,7 +43,40 @@ void RunnableObject::stop()
+std::pair<BasicPlan*, KeyMapping> RunnableObject::getParentPlanAndKeyMapping(RunningPlan* rp) const
+{
+    assert(rp->getParent()->getBasicPlan());
+    BasicPlan* parentPlan = rp->getParent()->getBasicPlan();
```
Base class (RunnableObject) having a reference to derived (BasicPlan) class is bad. Better to move mapping to the base object. Behaviour won't have children and we don't need to duplicate the mappings anyways.

christianhelm replied:
I understand the problem, but can you describe your suggested solution a bit more. The KeyMapping is associated with states and can't be in the RunnableObject.

christianhelm replied:
I have implemented a solution that makes use of RunningPlan

abhi1287 commented on alica_engine/include/engine/RunnableObject.h:
```diff
@@ -85,13 +90,14 @@ class RunnableObject
     std::atomic<RunningPlan*> _signalContext; // The running plan context when start() is called
     std::atomic<RunningPlan*> _execContext;   // The running plan context under which the behaviour is executing
     int64_t _activeRunJobId;
+    BlackboardBlueprint _blackboardBlueprint;
```
could be a const *

abhi1287 commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -52,44 +86,29 @@ void RunnableObject::start(RunningPlan* rp)
-            if (!planAttachment->setParameters(*parentPlan->getBlackboard(), *_blackboard)) {
-                std::cerr << "Setting parameters failed, supposedly as the context has already changed.  Plan will not be scheduled" << std::endl;
-                return;
+                _blackboard = std::make_shared<Blackboard>(_blackboardBlueprint);
```
This could be quite a heavy operation every time a behaviour / plan is activated. Please put a "TBD: optimize" comment here - can fix in a later PR.

abhi1287 commented on alica_engine/include/engine/blackboard/KeyMapping.h:
```diff
@@ -0,0 +1,26 @@
+    KeyMapping() = default;
+    virtual ~KeyMapping() = default;
+    void setInput(const Blackboard* parent_bb, Blackboard* child_bb) const;
+    void setOutput(Blackboard* parent_bb, const Blackboard* child_bb) const;
```
setInput  and setOutput method could be in the RunnableObject

abhi1287 commented on alica_engine/include/engine/model/Behaviour.h:
```diff
@@ -34,11 +35,12 @@ class Behaviour : public AbstractPlan
     const RuntimeCondition* getRuntimeCondition() const { return _runtimeCondition; }
     const PreCondition* getPreCondition() const { return _preCondition; }
     const PostCondition* getPostCondition() const { return _postCondition; }
+    BlackboardBlueprint getBlackboardBlueprint() const { return _blackboardBlueprint; }
```
return const ptr

abhi1287 commented on alica_engine/include/engine/model/Behaviour.h:
```diff
@@ -63,11 +65,11 @@ class Behaviour : public AbstractPlan
-     * If so, the corresponding autogenerated function will be called
-     * Otherwise, it will simply receive a copy of its parents Blackboard
+     * Specifies whether this behavior uses its parents' blackboard.
+     * If so, it will simply receive a copy of its parents Blackboard
```
it will simply receive a copy of  -> reference to

abhi1287 commented on alica_engine/include/engine/model/Plan.h:
```diff
@@ -54,7 +54,8 @@ class Plan : public AbstractPlan
 
-    bool getRequiresParameters() const { return _requiresParameters; }
+    bool getInheritBlackboard() const { return _inheritBlackboard; }
+    BlackboardBlueprint getBlackboardBlueprint() const { return _blackboardBlueprint; }
```
const *

abhi1287 commented on alica_engine/include/engine/model/Plan.h:
```diff
@@ -113,11 +115,11 @@ class Plan : public AbstractPlan
-     * If so, the corresponding autogenerated function will be called
-     * Otherwise, it will simply receive a copy of its parents Blackboard
+     * Specifies whether this plan inherits its parents Blackboard.
+     * If so, it will simply receive a copy of its parents Blackboard
```
reference

abhi1287 commented on alica_engine/include/engine/modelmanagement/factories/BlackboardFactory.h:
```diff
@@ -0,0 +1,19 @@
+namespace alica
+{
+class BlackboardBlueprint;
+class BlackboardFactory : public Factory
```
BlackboardBlueprintFactory ?

abhi1287 commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -42,7 +43,40 @@ void RunnableObject::stop()
+    }
+    ++_signalState;
+    if (rp->getParent()) {
+        if (!_inheritBlackboard) {
```
can club these 2 if conditions into 1

abhi1287 commented on alica_engine/src/engine/blackboard/KeyMapping.cpp:
```diff
@@ -0,0 +1,44 @@
+    for (const auto& [parentKey, childKey] : outputMapping) {
+        lockedParentBb.set(parentKey, lockedChildBb.get(childKey));
+        ALICA_DEBUG_MSG("passing " << childKey << " into " << parentKey);
+    }
```
We should have a more graceful handling of missing keys in blackboards. 
We can probably throw exception with a warning log that either of the blackboard is missing keys or can just throw a warning message and continue. 
I am inclined towards the 2nd option because stops could be triggered by interrupts before plans could put values to output variables.

abhi1287 commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -52,44 +86,29 @@ void RunnableObject::start(RunningPlan* rp)
+            keyMapping.setInput(parentPlan->getBlackboard().get(), _blackboard.get());
+            doInit();
+        };
+    } else if (_inheritBlackboard) {
```
could just be "else". we never want to schedule empty initCall

christianhelm committed:
Remove unused function

christianhelm committed:
Don't use lock at key mapping

christianhelm committed:
Improve blueprint generation and storage

christianhelm committed:
Use pointer for key mapping

christianhelm committed:
Add comments

christianhelm committed:
setInput and setOutput as member functions

christianhelm committed:
Fix comments

christianhelm committed:
Restructure if statements

christianhelm committed:
Handle missing blackboard values

christianhelm committed:
Test missing keys

christianhelm committed:
Merge branch 'rr_devel' into blackboard_json_2

christianhelm committed:
Improve dependencies

christianhelm committed:
Merge branch 'rr_devel' into blackboard_json_2

abhi1287 commented on alica_engine/src/engine/modelmanagement/factories/KeyMappingFactory.cpp:
```diff
@@ -0,0 +1,29 @@
+const KeyMapping* KeyMappingFactory::create(const YAML::Node& node)
+{
+
+    auto keyMapping = new KeyMapping();
```
this should be returned and stored in ConfAbstractPlanWrapper as a unique pointer.  By returning a const, no one would delete this memory when the engine terminates and this will show up as memory leak in all static analysis tools.

abhi1287 commented on alica_engine/src/engine/modelmanagement/factories/BlackboardBlueprintFactory.cpp:
```diff
@@ -0,0 +1,25 @@
+const BlackboardBlueprint* BlackboardBlueprintFactory::create(const YAML::Node& node)
+{
+
+    auto blackboard = new BlackboardBlueprint();
```
Same comment as keyMappingFactory. Return unique_ptr.

abhi1287 commented on alica_engine/include/engine/model/ConfAbstractPlanWrapper.h:
```diff
@@ -16,11 +19,13 @@ class ConfAbstractPlanWrapper : public AlicaElement
 
     const AbstractPlan* _abstractPlan;
     const Configuration* _configuration;
+    const KeyMapping* _keyMapping;
```
initialize with nullptr

abhi1287 commented on alica_engine/include/engine/RunningPlan.h:
```diff
@@ -137,6 +137,8 @@ class RunningPlan : public std::enable_shared_from_this<RunningPlan>
     BasicBehaviour* getBasicBehaviour() const { return _basicBehaviour; }
     BasicPlan* getBasicPlan() const { return _basicPlan; }
+    std::shared_ptr<Blackboard> getBlackboard() const;
+    const KeyMapping* getKeyMapping(int64_t wrapperId) const;
```
This still isn't clean - RunningPlan uses BasicPlan (RunnableObject)  which in turn uses RunningPlan. But that dependency existed already so lets do the cleanup in a different pr.

christianhelm replied:
I have created a new task for that https://www.wrike.com/open.htm?id=840847055

abhi1287 commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -149,4 +176,36 @@ void RunnableObject::traceInit(const std::string& type)
+            ALICA_DEBUG_MSG("passing " << parentKey << " into " << childKey);
+        } catch (std::exception& e) {
+            ALICA_WARNING_MSG("Blackboard error passing " << parentKey << " into " << childKey << ". " << e.what());
+        }
```
This and output would be really nice information to put onto trace. Please create a new pr with the trace change once this is merged.

christianhelm replied:
New task https://www.wrike.com/open.htm?id=840847670

abhi1287 commented on alica_engine/src/engine/blackboard/KeyMapping.cpp:
```diff
@@ -0,0 +1,35 @@
+    return _inputMapping;
+}
+
+KeyMapping::KeyMappingList KeyMapping::getOutputMapping() const
```
could be const&

christianhelm committed:
Store in unique ptr

christianhelm committed:
Init with nullptr

christianhelm committed:
Return reference

Dale-Koenig commented on alica_engine/include/engine/RunnableObject.h:
```diff
@@ -37,11 +39,13 @@ class RunnableObject
+    bool getInheritBlackboard() const { return _blackboardBlueprint == nullptr; };
+    void setBlackboardBlueprint(const BlackboardBlueprint* blackboard);
+    void stop(RunningPlan* rp);
+    void stop(); // Use only when shutdown engine
```
Just rename to `onShutdown` or something

Dale-Koenig commented on alica_engine/include/engine/RunnableObject.h:
```diff
@@ -37,11 +39,13 @@ class RunnableObject
-    bool getRequiresParameters() const { return _requiresParameters; }
-    void setRequiresParameters(bool requiresParameters) { _requiresParameters = requiresParameters; }
-    void stop();
+    bool getInheritBlackboard() const { return _blackboardBlueprint == nullptr; };
```
What about for master plan (doesn't inherit, but no blackboard blueprint)

christianhelm replied:
For a master plan this could return wrong results. Where the blackboard is used (RunnableObject::start, RunnableObject::stop) this is combined with a check if there is a parent.

Dale-Koenig commented on alica_engine/include/engine/model/Behaviour.h:
```diff
@@ -80,6 +76,11 @@ class Behaviour : public AbstractPlan
      */
     PostCondition* _postCondition;
+    /**
+     * If nullptr it will simply receive a reference to it's parents Blackboard
```
```suggestion
     * If nullptr it will simply receive a reference to its parent's Blackboard
```

christianhelm committed:
Update alica_engine/include/engine/model/Behaviour.h

Co-authored-by: Dale Koenig <dale.koenig@rapyuta-robotics.com>
